signer/src/crypto/derive_junction.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use codec::{Decode, Encode};
+
+/// The length of the junction identifier. Note that this is also referred to as the
+/// `CHAIN_CODE_LENGTH` in the context of Schnorrkel.
+pub const JUNCTION_ID_LEN: usize = 32;
+
+/// A since derivation junction description. It is the single parameter used when creating
+/// a new secret key from an existing secret key and, in the case of `SoftRaw` and `SoftIndex`
+/// a new public key from an existing public key.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug, Encode, Decode)]
+pub enum DeriveJunction {
+    /// Soft (vanilla) derivation. Public keys have a correspondent derivation.
+    Soft([u8; JUNCTION_ID_LEN]),
+    /// Hard ("hardened") derivation. Public keys do not have a correspondent derivation.
+    Hard([u8; JUNCTION_ID_LEN]),
+}
+
+impl DeriveJunction {
+    /// Consume self to return a soft derive junction with the same chain code.
+    pub fn soften(self) -> Self {
+        DeriveJunction::Soft(self.unwrap_inner())
+    }
+
+    /// Consume self to return a hard derive junction with the same chain code.
+    pub fn harden(self) -> Self {
+        DeriveJunction::Hard(self.unwrap_inner())
+    }
+
+    /// Create a new soft (vanilla) DeriveJunction from a given, encodable, value.
+    ///
+    /// If you need a hard junction, use `hard()`.
+    pub fn soft<T: Encode>(index: T) -> Self {
+        let mut cc: [u8; JUNCTION_ID_LEN] = Default::default();
+        index.using_encoded(|data| {
+            if data.len() > JUNCTION_ID_LEN {
+                cc.copy_from_slice(&sp_core_hashing::blake2_256(data));
+            } else {
+                cc[0..data.len()].copy_from_slice(data);
+            }
+        });
+        DeriveJunction::Soft(cc)
+    }
+
+    /// Create a new hard (hardened) DeriveJunction from a given, encodable, value.
+    ///
+    /// If you need a soft junction, use `soft()`.
+    pub fn hard<T: Encode>(index: T) -> Self {
+        Self::soft(index).harden()
+    }
+
+    /// Consume self to return the chain code.
+    pub fn unwrap_inner(self) -> [u8; JUNCTION_ID_LEN] {
```

Comment:
I would prefer `into_inner` instead of `unwrap_inner` because I regard `unwrap` as something fallible but maybe just me :P

Reply:
Agreed! This is basically copied verbatim from sp_core so I hadn't really paid it much attention :)